### ISSUE TITLE
Unable to use Flux in function signature

### ISSUE THREAD

**Issue by trisberg** [projectriff/riff-buildpack#7]
I have a boot app with a "<Flux<String>, Flux<String>>" signature and it works running from command line and with the older jar invoker build with Kaniko. It doesn't work when invoking the function and I get `{"timestamp":"Oct 10, 2018 5:48:04 PM","status":500,"error":"Internal Server Error","message":"java.lang.String cannot be cast to reactor.core.publisher.Flux","path":"/"}`.

Test app:

```
riff function create java cap \
  --git-repo https://github.com/trisberg/capitalize-boot.git \
  --image gcr.io/cf-spring-funkytown/capitalize-boot \
  --git-revision flux \
  --verbose
riff service invoke cap -- -w '\n' -H 'Content-Type: text/plain' -d 'hello there, how are you?'
riff service delete cap
```

**Comment by trisberg** [comment]
Source is here in the "flux" branch: https://github.com/trisberg/capitalize-boot/tree/flux

**Comment by trisberg** [comment]
Did some more troubleshooting and if I extract the invoker jar I can start the app using 
```
java -cp '/workspace/io.projectriff.riff/riff-invoker-java:/workspace/io.projectriff.riff/riff-invoker-java/BOOT-INF/classes:/workspace/io.projectriff.riff/riff-invoker-java/BOOT-INF/lib/*' $JAVA_OPTS io.projectriff.invoker.JavaFunctionInvokerApplication --function.uri='file:///workspace/app?handler=capitalize'
```
and everything works fine

**Comment by nebhale** [comment]
I'm not precisely sure what the problem is here, but looking at the command line, I don't think it's anything that should be a production value.  Bypassing Boot's class loading using `-cp` like that is ripe for future problems as they change and evolve the class loading strategies of Boot.

**Comment by dsyer** [comment]
> Bypassing Boot's class loading using -cp like that is ripe for future problems as they change and evolve the class loading strategies of Boot.

I don't think so. Boot's class loader is only really useful for running jar files. This is an exploded archive. It will always run better from its own main class. If it helps we can use a non-fat archive from the invoker - we control the whole cycle here, so there's nothing that says we need to use a Boot fat jar.

**Comment by nebhale** [comment]
Well, as we've seen elsewhere, the way that Boot creates its classloader isn't necessarily the way a classloader would be created at the command line.  Iteration-order being the biggest hangup (quite a number of CF support issues around this) but I believe the classloader hierarchy would be different as well.

/cc @wilkinsona @philwebb

**Comment by wilkinsona** [comment]
Seeing the full stack would be interesting.

It’s an unusual cause of a class cast exception, but it looks like a classpath ordering problem to me. When running as a fat jar, it’ll use the order of the entries in the jar as determined by Maven/Gradle at build time. Once exploded you’ll get whatever ordering your file system and shell give you.

The hierarchy will be different too, but I’m not sure that it’ll be functionally different.

**Comment by dsyer** [comment]
The full stack wouldn’t tell you what you think it might. The problem is to do with the way the Riff function invoker creates its own class loader. It assumes that it is created and run from an exploded archive because that’s what gives you the fastest cold start. We could fix it there (and I probably will just for consistency), but to get the best runtime performance we should allow it to start from `java -cp ...` if it wants to, and we should implement this buildpack in that way.

**Comment by wilkinsona** [comment]
Intriguing. I’m looking forward to learning how the creation of the class loader leads to the ClassCastException.

I think https://github.com/spring-projects/spring-boot/issues/9128 should be kept in mind. Once that’s implemented, using Boot’s launcher with an exploded archive will give you the same ordering of the classpath as you’d get with a packaged archive. At that point I think it would be a mistake to use a manually constructed classpath as I being proposed here. With one eye on the future, I’d explode the archive but use Boot’s launcher today.

**Comment by dsyer** [comment]
I don't really want this to be an app where ordering of the classpath matters. We control all the dependencies, so I don't see why it would. But thanks for the heads up about the new feature.

**Comment by wilkinsona** [comment]
That's a nice position to be in. Given that you control all the dependencies, it should be possible for ordering to be unimportant. I guess this issue will figure out why, apparently, the ordering currently is important.
